### PR TITLE
[Sprint 9 / Enhancement] Refactor Keisan

### PR DESCRIPTION
**Pull request by maroqijalil**
## Jira Link: 

https://ichiro-its.atlassian.net/browse/PI-156

## Description

Adding 4 by 4 Matrix Transformation operation for 3D Transformation needed. 

## Type of Change

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [x] New unit tests added.
- [ ] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: include/keisan/matrix/matrix_3d.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace keisan
+{
+
+class Point3;
+
+class Matrix3D
+{
+public:
+  static Matrix<4, 4> identity();
```
Comment (threeal) [inline include/keisan/matrix/matrix_3d.hpp:37]: can we create a template here? i suggest to put it in `matrix.hpp` too as it's more general
```
template<N>
static Matrix<N, N> identity();
```

File: include/keisan/matrix/matrix_3d.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  static bool inverse(std::shared_ptr<Matrix<4, 4>> matrix);
```
Comment (threeal) [inline include/keisan/matrix/matrix_3d.hpp:39]: the same with `identity()`

File: include/keisan/matrix/matrix_3d.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  static Matrix<4, 4> transformation_of(const Point3 & point, const Vector<3> & angle);
```
Comment (threeal) [inline include/keisan/matrix/matrix_3d.hpp:41]: if we want to work with matrix transformation, i suggest to use these format instead, e.g. :
```c++
Matrix<4, 4> translation_matrix(const Point3 & point);
Matrix<3, 3> translation_matrix(const Point2 & point);
Matrix<3, 3> rotation_matrix(const Angle & angle);
// ...
```

Comment (threeal) [inline include/keisan/matrix/matrix_3d.hpp:41]: use `Euler` or `Quaternion` instead of `Vector<3>` to represent angle position in 3d space

**Review by threeal — CHANGES_REQUESTED** [review]

File: include/keisan/matrix/matrix_3d.hpp (hunk 1 above)
Comment (maroqijalil) [reply include/keisan/matrix/matrix_3d.hpp:37]: Maybe, it can be

File: include/keisan/matrix/matrix_3d.hpp (hunk 2 above)
Comment (maroqijalil) [reply include/keisan/matrix/matrix_3d.hpp:39]: Actually, I created this helper class was just for helping `aruku` package to perform its inverse kinematic method. That is the reason why I created this and named this class as `Matrix3D`. I copied from the previous class on our framework with several changes.
Okay, probably I was wrong, I will set this PR as draft and define this class at `aruku` package instead. To make this package has a good structure and functionality, I think we can make specific task to add operations such these as general as possible. It's all about the time

File: include/keisan/matrix/matrix_3d.hpp (hunk 3 above)
Comment (maroqijalil) [reply include/keisan/matrix/matrix_3d.hpp:41]: I will revise the method in `aruku` package.

Comment (maroqijalil) [reply include/keisan/matrix/matrix_3d.hpp:41]: I will revise the method in `aruku` package.

**Comment by maroqijalil** [comment]
Maybe any suggestion before I move to `aruku` package?

**Comment by threeal** [comment]
> Maybe any suggestion before I move to `aruku` package?

i think it's better to be put in here, `identity()` should be easy to implement, there was implementation for that on `SquareMatrix`, but it was removed. As for `inverse()` idk how hard it's going to be, it's okay though for now to implement that only for 4x4 matrix. btw, i think you could do this, make it template but for now, the implementation is only available for 4x4 matrix:

```c++
template<size_t N>
bool Matrix<N, N>::inverse()
{
  static_assert(N != 4, "only available for 4x4 matrix");
  // implementation
}
```

and let's put `identity()` as static function of `Matrix`, and put `inverse()` as a member function of `Matrix`.

**Comment by maroqijalil** [comment]
I've moved those to `matrix.hpp`. For the rest, the specific task in regard to provide complete functionality of `inverse` method will be added.

File: include/keisan/matrix/matrix.hpp [hunk 4]
```diff
@@ -74,10 +75,30 @@ class Matrix
   double * operator[](size_t pos);
   const double * operator[](size_t pos) const;
 
+  bool inverse();
+
 private:
   double data[M * N];
 };
 
+struct Point3;
```
Comment (threeal) [inline include/keisan/matrix/matrix.hpp:84]: i suggest to put forward declaration (`struct x;`) on top most of this file, right after include declarations (`#include <x>`).

File: include/keisan/matrix/matrix.impl.hpp [hunk 5]
```diff
@@ -93,6 +93,24 @@ Matrix<M, N> Matrix<M, N>::zero()
...
+
+  Matrix<N, N> matrix;
+  for (size_t i = 0; i < N; ++i) {
+    for (size_t j = 0; j < N; ++j) {
+      matrix[i][j] = (i == j) ? 1.0 : 0.0;
+    }
+  }
+
+  return matrix;
+}
```
Comment (threeal) [inline include/keisan/matrix/matrix.impl.hpp:112]: i suggest to return `Matrix<M, N>` instead, even if in the end `M` will always equal to `N`.

**Review by threeal — APPROVED** [review]
Looks good though 👍

**Pull request merged**